Implement block breaking animation

satoshinm opened the pull request:
- [x] Add block break animation message and codec
- [x] Broadcast message to nearby clients
- [x] Calculate block break stage when breaking
- [x] Clear block break animation when stopped
- ~~factoring in tool effectiveness vs hardness~~ - out of scope of this pull request
- ~~BlockCrackEvent API event? (fire in between BlockDamageEvent and BlockBreakEvent)~~ - maybe later

turt2live commented:
@SpaceManiac This pull request is pending further review by yourself. Please take the time, if possible, to do so. Thanks!

satoshinm committed:
Add block break animation message and codec

satoshinm commented:
Added the 0x08 codec, but that's the easy part, I'll have to dig into the code further to see how to send this message to nearby clients appropriately (or if anyone wants to lend me some pointers how this would be done?). Note this message allows clients to see the block break stages when other players are digging blocks.

aramperes commented:
You can use something like [this](https://github.com/GlowstoneMC/Glowstone/blob/master/src/main/java/net/glowstone/block/entity/BlockEntity.java#L43) to update in range.

aramperes commented on src/main/java/net/glowstone/net/message/play/game/BlockBreakAnimationMessage.java:
```diff
@@ -0,0 +1,10 @@
+@Data
+public class BlockBreakAnimationMessage implements Message {
+
+    private final int entityID, x, y, z, destroyStage;
```
For the sake of continuity with other entity Message classes, change `entityID` to `id`.

satoshinm replied:
Changed to id

aramperes commented on src/main/java/net/glowstone/net/codec/play/game/BlockBreakAnimationCodec.java:
```diff
@@ -0,0 +1,29 @@
+        return buf;
+    }
+}
+
```
Extra whitespace.

mastercoms replied:
Is that not a newline at the end of file?

satoshinm replied:
It's an extra newline, I'll remove it *update: removed*

mastercoms commented on src/main/java/net/glowstone/net/message/play/game/BlockBreakAnimationMessage.java:
```diff
@@ -0,0 +1,10 @@
+public class BlockBreakAnimationMessage implements Message {
+
+    private final int entityID, x, y, z, destroyStage;
+}
```
Add a newline here please.

satoshinm replied:
Added newline after the variable declarations

mastercoms commented:
Ah, I thought GitHub was supposed to show newlines.

satoshinm committed:
Broadcast block break animation when player starts digging

satoshinm commented:
<img width="966" alt="screen shot 2017-05-15 at 9 44 07 pm" src="https://cloud.githubusercontent.com/assets/26856618/26090134/c11a8204-39b7-11e7-8d15-fd2a505bf6da.png">

Added an initial block break animation broadcast, other clients can now see the initial block break stage. However, the digging handler isn't continuously called, only when digging starts/stops. I think to implement this properly, the server would need to know each blocks' hardness to calculate the stages, then increment accordingly. Looking at another open source server (Cuberite) it seems they haven't implemented this part yet either: https://github.com/cuberite/cuberite/blob/01d9beee9ee60aa1ead52f7796c762a8372b9ca4/src/ClientHandle.cpp#L1223-L1231 "TODO: Send animation packets even without receiving any other packets" edit: actually, they update the stage in a "client tick" method: https://github.com/cuberite/cuberite/blob/01d9beee9ee60aa1ead52f7796c762a8372b9ca4/src/ClientHandle.cpp#L2134-L2147 - does Glowstone have something equivalent to this?

mastercoms commented:
Yeah but I don't think it is good to have it in a client tick. Digging is caused by the packet, and so any tasks should off shoot from there. 

About hardness, that info is stored in the material values collection.

satoshinm commented:
> Yeah but I don't think it is good to have it in a client tick. Digging is caused by the packet, and so any tasks should off shoot from there.

Start an asynchronous task if `message.getState() == DiggingMessage.START_DIGGING`, sending increasingly block break destroy stage packets at an interval, then cancel the task when the block breaks or digging stops (`message.getState() == DiggingMessage.FINISH_DIGGING`)?

[materialValues.yml](https://github.com/GlowstoneMC/Glowstone/blob/77e30f78e4e323ae61a999c7dad3075731985f77/src/main/resources/builtin/materialValues.yml) has a `hardness` field but only values used are: 1 (default), -1 (bedrock), 0 (soft blocks, plants etc.), and 50 (bedrock).

Hardness values to populate the yaml can be found at: http://minecraft.gamepedia.com/Breaking#Blocks_by_hardness

mastercoms commented:
This PR would probably depend on [#379](https://github.com/GlowstoneMC/Glowstone/pull/379/files#diff-9f83ce130d9b3bff618cb81945f17116R14), then.

About the async task, I think it could work like that.

satoshinm committed:
Add material hardness from http://minecraft.gamepedia.com/Breaking#Blocks_by_hardness

satoshinm commented:
Hm, didn't see https://github.com/GlowstoneMC/Glowstone/pull/379/files#diff-9f83ce130d9b3bff618cb81945f17116R14 .. are the hardnesses moving to BuiltinMaterialValueManager.java out of materialValues.yml? Went ahead and added to .yml before I saw this, http://minecraft.gamepedia.com/Breaking#Blocks_by_hardness has all the hardness values for reference except skulls but http://minecraft.gamepedia.com/Mob_head says skull hardness = 1.

mastercoms commented:
Yeah they're being moved to the class so it can better support custom materials and programmable modifications of vanilla materials.

However, your latest commit should be fine. It removes the need to wait for #379 to be merged first.

satoshinm committed:
Move block break animation broadcasting to GlowPlayer#setDigging()

satoshinm committed:
Handle digging canceled, stops before finished

satoshinm committed:
Send block break stages based on hardness

satoshinm committed:
Don't send block break animation to the player who caused it

satoshinm commented:
Addressed the requested changes but they still show as changes requested, how can I mark them addressed?

Implemented the block break animation stages, used player pulse() seemed to make the most sense there. Added cancel digging support. This is functional now however, I'm finding the hardness differs from what the player sees:

<img width="1560" alt="screen shot 2017-05-21 at 8 06 07 pm" src="https://cloud.githubusercontent.com/assets/26856618/26291130/17cad768-3e61-11e7-8cd3-97072a8017a2.png">

left = the player digging, right = the block break animation from another client (too fast).

It isn't taking into account the tool/effectiveness versus hardness. From http://minecraft.gamepedia.com/Breaking#Blocks_by_hardness - stone "hardness" is 1.5 seconds, but when mined with a bare hand it is 7.5 seconds. The equation is described in http://minecraft.gamepedia.com/Breaking#Speed:

> The base time in seconds is the block's hardness multiplied by 1.5 if the player can harvest the block with the current tool, or multiplied by 5 if the player cannot.

mastercoms commented:
The reviewer has to approve changes. Nothing you need to do.

Due to the complexity of this issue, I believe it is best to leave it for another PR. Initial block break animation support is sufficient, in my opinion.

satoshinm committed:
Update comments

satoshinm committed:
Add 5x multiplier to block hardness

satoshinm commented:
Agreed, sounds good. I changed it for now to use the correct hardness for if the block is mined with bare hands and is not dropped, for example, hands on stone -- the animations now match. The consequences of incorrect block break animation are purely cosmetic, tool effectiveness can be factored in later, I think this is good enough for now.

aramperes commented on src/main/java/net/glowstone/entity/GlowPlayer.java:
```diff
@@ -2826,10 +2846,47 @@ public GlowBlock getDigging() {
+        // for now, assuming hands are used and the block is not dropped
+        hardness *= 5;
+
+        long duration = System.currentTimeMillis() - diggingStarted;
```
Wouldn't it be better to base this off world thread ticking, instead of system time in milliseconds?

satoshinm replied:
Good point, will fix *update*: changed to count ticks

satoshinm committed:
Change digging to count ticks instead of using system time

mastercoms commented on src/main/java/net/glowstone/entity/GlowPlayer.java:
```diff
@@ -2826,10 +2846,47 @@ public GlowBlock getDigging() {
     }
 
+    private void broadcastBlockBreakAnimation(GlowBlock block, int destroyStage) {
+        if (block == null) return;
```
Wouldn't this work better as a precondition?

satoshinm replied:
Agreed, changed to move the check outside the function - it should only be called with non-null block now

mastercoms commented on src/main/java/net/glowstone/entity/GlowPlayer.java:
```diff
@@ -2826,10 +2846,47 @@ public GlowBlock getDigging() {
+
+        GlowChunk.Key key = new GlowChunk.Key(block.getChunk().getX(), block.getChunk().getZ());
+        // TODO: should canSeeChunk filtering take place here, in GlowPlayer, or both?
+        block.getWorld().getRawPlayers().stream().filter(player1 -> player1.canSeeChunk(key)).forEach(player2 -> {
```
Could you explain to me what this code is doing?

satoshinm replied:
It is getting the chunk of the block and filtering the list of players to those who can see the chunk, so it only sends the block breaking animation to the right players. I wasn't familiar with the Streams API and Java lambdas but it seems to be a common pattern in Glowstone:

```
src/main/java/net/glowstone/block/entity/BlockEntity.java
45:        block.getWorld().getRawPlayers().stream().filter(player -> player.canSeeChunk(key)).forEach(this::update);

src/main/java/net/glowstone/block/state/GlowNoteBlock.java
189:        getWorld().getRawPlayers().stream().filter(player -> player.canSeeChunk(key)).forEach(player -> player.playNote(location, instrument, note));

src/main/java/net/glowstone/entity/GlowEntity.java
1024:        world.getRawPlayers().stream().filter(player -> player.canSeeEntity(this)).forEach(player -> player.getSession().send(message));

src/main/java/net/glowstone/entity/objects/GlowItem.java
71:            world.getRawPlayers().stream().filter(other -> other.canSeeEntity(this)).forEach(other -> other.getSession().send(message));
```

but the players are filtered twice for chunk visibility, both here and in sendBlockBreakAnimation:

```java
    public void sendBlockBreakAnimation(BlockBreakAnimationMessage message) {
        // only send message if the chunk is within visible range
        Key key = new Key(message.getX() >> 4, message.getZ() >> 4);
        if (canSeeChunk(key)) {
            afterBlockChanges.add(message);
        }
    }
```

leaning towards removing this second check and only performing the filtering in broadcastBlockBreakAnimation **update**: changed to filter only once

mastercoms replied:
Ok. Why did you name your variables player1 and player2? Can't they both be player?

Also you have an if statement within the for-each lambda that could be part of the filter predicate.

satoshinm replied:
Made both changes, also removed unnecessary Player parameter

satoshinm committed:
Move null check outside broadcastBlockBreakAnimation() as a precondition

satoshinm committed:
Move sendBlockBreakAnimation() down nearby broadcastBlockBreakAnimation()

satoshinm committed:
Check for player chunk visibility only once

mastercoms commented:
@satoshinm GitHub is wrongly hiding my comment here: https://github.com/GlowstoneMC/Glowstone/pull/473#discussion_r117816322

I'm not sure if you saw it because of this.

satoshinm committed:
Simplify broadcastBlockBreakAnimation() lambdas

satoshinm commented:
@mastercoms thanks didn't see it, will fix *update*: made the changes

mastercoms commented:
Alright, looks good. Thank you for your contribution to Glowstone.

mastercoms commented:
By the way, did you see our invite at https://github.com/GlowstoneMC?

(As per our contribution policy: https://github.com/GlowstoneMC/Glowstone/blob/master/.github/CONTRIBUTING.md)
